### PR TITLE
kernel: Add k_event_set_masked primitive

### PR DESCRIPTION
**Pull request by ajackson-xlnx**
This is a possible, partial fix to issue#46117 which allows a consumer of events to avoid a race with a producer (or producers).

### CODE REVIEW

**Review by peter-mitsis — CHANGES_REQUESTED** [review]
I really like the ideas you are proposing here. There are just a few loose ends that should be cleaned up a little.

File: kernel/events.c [hunk 1]
```diff
@@ -95,12 +95,10 @@ static void k_event_post_internal(struct k_event *event, uint32_t events,
 	key = k_spin_lock(&event->lock);
 
 	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_event, post, event, events,
-					accumulate);
-
-	if (accumulate) {
-		events |= event->events;
-	}
+					events_mask);
```
[the remark this answers is not in the document]
Comment (ajackson-xlnx) [reply kernel/events.c:98]: Done

File: kernel/events.c [hunk 2]
```diff
@@ -145,12 +143,12 @@ static void k_event_post_internal(struct k_event *event, uint32_t events,
 	z_reschedule(&event->lock, key);
 
 	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_event, post, event, events,
-				       accumulate);
+				       events_mask);
```
[the remark this answers is not in the document]
Comment (ajackson-xlnx) [reply kernel/events.c:146]: Done

File: include/zephyr/kernel.h [hunk 3]
```diff
@@ -2128,6 +2128,24 @@ __syscall void k_event_post(struct k_event *event, uint32_t events);
...
+ * @brief Set or clear the events in an event object
+ *
+ * This routine sets the events stored in event object to the specified value.
+ * All tasks waiting on the event object @a event whose waiting conditions
+ * become met by this immediately unpend. Unlike @ref k_event_set, this routine
+ * allows specific event bits to be set and cleared as determined by the mask.
+ *
+ * Setting differs from posting in that set events replace the current set of
+ * events tracked by the event object.
+ *
```
[the remark this answers is not in the document]
Comment (ajackson-xlnx) [reply include/zephyr/kernel.h]: Done

File: include/zephyr/tracing/tracing.h [hunk 4]
```diff
@@ -1870,17 +1870,17 @@
  * @brief Trace posting of an Event call entry
  * @param event Event object
  * @param events Set of posted events
- * @param accumulate True if events accumulate, false otherwise
+ * @param events_mask Mask to apply against posted events
  */
-#define sys_port_trace_k_event_post_enter(event, events, accumulate)
+#define sys_port_trace_k_event_post_enter(event, events, events_mask)
```
Comment (peter-mitsis) [inline include/zephyr/tracing/tracing.h:1875]: This basis of this comment also applies to the sys_port_trace_k_event_post_exit() change.

Almost there. There are a couple of other references to sys_port_trace_k_event_post_enter() has some references in ...

subsys/tracing/test/tracing_test.h
subsys/tracing/user/tracing_user.h
subsys/tracing/ctf/tracing_ctf.h

That **accumulate** parameter in those files will also need to be updated to **events_mask** as well.

**Review by peter-mitsis — CHANGES_REQUESTED** [review]
Just one last set of changes to go, and from my perspective it should be good.

File: include/zephyr/tracing/tracing.h (hunk 4 above)
Comment (ajackson-xlnx) [reply include/zephyr/tracing/tracing.h:1875]: Sorry I missed those: thank you for your patience (again)

**Review by peter-mitsis — APPROVED** [review]
This looks good to me.

**Review by andyross — APPROVED** [review]
Very clean.

**Pull request merged**